### PR TITLE
New menu system

### PR DESCRIPTION
**Pull request by psav**
Major changes to the menu library here, in order to support 5.6s new side based nav. This new nav not only is composed differently with new elements, but is triple layered, instead of being only double layered.

Summary of changes:

 * Added a pending UINavigate() class to overide the original, this is to make the ui_navigate module a class, and not a singleton.
 * Fixes to various object model files to get them to import in the correct way
 * Branches are now not added on import, instead they are pushed into a deferral stack which is invoked the first time you call force_navigate(). This now makes force_navigate() the _ONLY_ way to
   navigate and this function may very well move inside the menu module soon.
 * The menu module has been reshaped to give a class based Menu() object. This is to allow multiple Menu objects to be created, removing yet another singleton.
 * Currently the new Menu() object is exposed as a module level "nav" object this was to keep inline with current code and minimise non-related changes in this PR.

### CODE REVIEW

File: cfme/web_ui/menu.py [hunk 1]
```diff
@@ -140,8 +239,35 @@ def is_page_active(toplevel, secondlevel=None):
     return True
 
 
-def nav_to_fn(toplevel, secondlevel=None, reset_action=None, _final=False):
+def _try_nav(el):
+    try:
+        href = sel.get_attribute(el, 'href')
+        sel.execute_script('document.location.href="{}"'.format(href))
+    except NoSuchElementException:
+        raise
```
Comment (mfalesni) [inline cfme/web_ui/menu.py]: Does this make sense?

**Comment by dajoRH** [comment]
Lint report for commit 8f0ef5d0f1cff42989efb720f3aa9ee48aacee06:

`cfme/web_ui/menu.py`:
- :grey_exclamation: Line 270:9: **T001** *print statement found.*
- :red_circle: Line 458:1: **F811** *redefinition of unused 'branch_convert' from line 442*
- :grey_exclamation: Line 461:9: **T001** *print statement found.*
- :bangbang: Line 471:101: **E501** *line too long (105 > 100 characters)*
- :bangbang: Line 472:101: **E501** *line too long (107 > 100 characters)*
- :bangbang: Line 482:101: **E501** *line too long (101 > 100 characters)*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit a609abfdfa0faa14e463437ae8a6bed8cc7460b3:

`cfme/web_ui/menu.py`:
- :grey_exclamation: Line 301:13: **T001** *print statement found.*
- :grey_exclamation: Line 447:13: **T001** *print statement found.*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit d7874c0ea547f968b101af9c51b0e2195d8d68fc:

Everything seems all right :smile: .
*CFME QE Bot*

File: cfme/web_ui/menu.py [hunk 2]
```diff
@@ -1,188 +1,284 @@
...
-            try:
-                href = sel.get_attribute(el, 'href')
-                sel.execute_script('document.location.href="{}"'.format(href))
-            except NoSuchElementException:
-                raise
+    @staticmethod
+    def _try_nav(el):
+        try:
+            href = sel.get_attribute(el, 'href')
+            sel.execute_script('document.location.href="{}"'.format(href))
```
Comment (mfalesni) [inline cfme/web_ui/menu.py]: This deserves an improvement :)

```python
sel.execute_script('document.location.href = arguments[0];', href)
```

**Comment by dajoRH** [comment]
Lint report for commit d3a4f06fcf009abc3118d814254a2550dd778dca:

`ui_navigate.py`:
- :grey_exclamation: Line 84:54: **S001** *found %s formatter*
- :grey_exclamation: Line 92:56: **S001** *found %s formatter*
- :grey_exclamation: Line 122:73: **S001** *found %s formatter*
- :grey_exclamation: Line 127:53: **S001** *found %s formatter*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 1354df47761afab6435c6deb810d8ed2c8110641:

`ui_navigate.py`:
- :grey_exclamation: Line 84:54: **S001** *found %s formatter*
- :grey_exclamation: Line 92:56: **S001** *found %s formatter*
- :grey_exclamation: Line 122:73: **S001** *found %s formatter*
- :grey_exclamation: Line 127:53: **S001** *found %s formatter*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 5d8b5b9d0cab772f33ba609c2a34332bdc31723a:

`ui_navigate.py`:
- :grey_exclamation: Line 84:54: **S001** *found %s formatter*
- :grey_exclamation: Line 92:56: **S001** *found %s formatter*
- :grey_exclamation: Line 122:73: **S001** *found %s formatter*
- :grey_exclamation: Line 127:53: **S001** *found %s formatter*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 1a00dc33261fb8678a94cbd6486902c0443bd836:

`ui_navigate.py`:
- :grey_exclamation: Line 84:54: **S001** *found %s formatter*
- :grey_exclamation: Line 92:56: **S001** *found %s formatter*
- :grey_exclamation: Line 122:73: **S001** *found %s formatter*
- :grey_exclamation: Line 127:53: **S001** *found %s formatter*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit ffe9400e5c43a0f193e218b8e63c8e10752d7402:

`cfme/web_ui/menu.py`:
- :bangbang: Line 27:5: **E303** *too many blank lines (2)*

`ui_navigate.py`:
- :grey_exclamation: Line 84:54: **S001** *found %s formatter*
- :grey_exclamation: Line 92:56: **S001** *found %s formatter*
- :grey_exclamation: Line 122:73: **S001** *found %s formatter*
- :grey_exclamation: Line 127:53: **S001** *found %s formatter*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit f4d153e08e5b108a89e5fde8a12c7f3399ae81f2:

`ui_navigate.py`:
- :grey_exclamation: Line 84:54: **S001** *found %s formatter*
- :grey_exclamation: Line 92:56: **S001** *found %s formatter*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 2af68dbbf4a472219ea12c55fe8e1b74c3978bc4:

`ui_navigate.py`:
- :grey_exclamation: Line 84:54: **S001** *found %s formatter*
- :grey_exclamation: Line 92:56: **S001** *found %s formatter*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit fd4610bccdf2469bc885c7f8d408eceb53ce9327:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 1a3e7c55c927e45bef0b0114d992f22605bff0c0:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit b9cff68a8f1f72e056060693612cf6d474281c0c:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 1e4515e8fc01b217e40b450bba801cd166696757:

Everything seems all right :smile: .
*CFME QE Bot*

**Pull request merged**